clang/lib/Sema/SemaCoroutine.cpp

```diff
@@ -1098,12 +1098,39 @@ static TypeSourceInfo *getTypeSourceInfoForStdAlignValT(Sema &S,
   return S.Context.getTrivialTypeSourceInfo(StdAlignValDecl);
 }
 
+// When searching for custom allocators on the PromiseType we want to
+// warn that we will ignore type aware allocators.
+static bool DiagnoseTypeAwareAllocators(Sema &S, SourceLocation Loc,
+                                        unsigned DiagnosticID,
+                                        DeclarationName Name,
+                                        QualType PromiseType) {
+  assert(PromiseType->isRecordType());
+
+  LookupResult R(S, Name, Loc, Sema::LookupOrdinaryName);
+  S.LookupQualifiedName(R, PromiseType->getAsCXXRecordDecl());
+  bool HaveIssuedWarning = false;
+  for (auto Decl : R) {
+    if (!Decl->getAsFunction()->isTypeAwareOperatorNewOrDelete())
+      continue;
+    if (!HaveIssuedWarning) {
+      S.Diag(Loc, DiagnosticID) << Name;
+      HaveIssuedWarning = true;
+    }
+    S.Diag(Decl->getLocation(), diag::note_type_aware_operator_declared)
+        << /* isTypeAware */ 1 << Decl << Decl->getDeclContext();
```

Comment:
I am a bit confused about this one.
For reference, the diagnostic format for this note is: `%select{non-|}0type aware %1 declared here in %2`

The note points to the `Decl`'s location, but the diagnostic says 'here in %2' which is the Decl's semantic parent.

Should you adjust the note's location?

Is it possible the semantic vs lexical context distinction could matter here, for example can these operators be declared as friends?

Small nit for consistency:
```suggestion
    S.Diag(Decl->getLocation(), diag::note_type_aware_operator_declared)
        << /*isTypeAware=*/1 << Decl << Decl->getDeclContext();
```

Reply:
My rationale is that while we are pointing to the operator decl directly, depending on the class (size or complexity hierarchy) having the declared scope immediately in the diagnostic note seemed useful? I can remove it if we think it's really unnecessary.

For the style nit I'll do a quick search for any similar patterns of missing the = in the commnet

Reply:
I think this is fine, you just have to be careful if the semantic context is the always the same as the lexical context.

Reply:
Mercifully for P2719 the semantically relevant location is the declaration scope is the scope of the original function declaration and there's now a test that:
```cpp
struct A { operator new ...; operator delete ...; }
struct B { operator new ...; operator delete ...; }
struct C : A, B {
   using A::operator new;
   using B::operator delete;
};
```
is detected as invalid due to new and delete coming from different scopes.

Reply:
But my original question, can these operators be declared as friends?

For a friend function, the semantic context and lexical context are different.

A friend function can be lexically inside a class, but semantically in the TU or namespace.

I think in this case you may want getLexicalDeclContext instead of getDeclContext.

Reply:
I believe we've established that getDeclContext is correct, and that friend, etc decls can't be used to pull new/delete around into unexpected scopes.

(Strictly, consider a hypothetical where you could pull new/delete in from other scopes, e.g.
```cpp
class SomeClass {
   operator new = yoinked from some other context;
   operator delete = yoinked from a different context;
}
```
We would not want to be reporting the lexical context (`SomeClass`) because the error is new and delete coming from different decl contexts, e.g. the fact that new and delete decls in SomeClass have the same lexical context is not relevant, the actual functions are from different contexts that would be a mismatch.